File: 983. Minimum Cost For Tickets/983. Minimum Cost For Tickets bottom-up.py
from functools import lru_cache
from typing import List


class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dp = [0]*(days[-1]+1)
        for i in range(len(dp)):
            if i in days:
                dp[i] = min(dp[max(0, i-1)]+costs[0],
                            dp[max(0, i-7)]+costs[1], dp[max(0, i-30)]+costs[2])
            else:
                dp[i] = dp[i-1]
        return dp[-1]


s = Solution()
days = [1, 4, 6, 7, 8, 20]
costs = [2, 7, 15]
print(s.mincostTickets(days, costs))

File: 983. Minimum Cost For Tickets/983. Minimum Cost For Tickets bottom-up.java
import java.util.Arrays;

class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        int[] dp = new int[days[days.length - 1] + 1];
        for (int i = 0; i < dp.length; i++) {
            if (Arrays.binarySearch(days, i) >= 0) {
                dp[i] = Math.min(costs[0] + dp[Math.max(0, i - 1)],
                        Math.min(costs[1] + dp[Math.max(0, i - 7)],
                                costs[2] + dp[Math.max(0, i - 30)]));
            } else {
                dp[i] = dp[Math.max(i - 1, 0)];
            }
        }
        return dp[days[days.length - 1]];
    }
}